apps/studio/src/lib/db/clients/clickhouse/ClickHouseCursor.ts

```diff
@@ -0,0 +1,66 @@
+import { BeeCursor } from "../../models";
+import rawlog from "electron-log";
+import type { ClickHouseClient, Row, StreamReadable } from "@clickhouse/client";
+import { uuidv4 } from "@/lib/uuid";
+
+interface CursorOptions {
+  query: string;
+  params: Record<string, any>;
+  client: ClickHouseClient;
+  chunkSize: number;
+}
+
+const log = rawlog.scope("clickhouse/cursor");
+
+export class ClickHouseCursor extends BeeCursor {
+  private readonly options: CursorOptions;
+  private client: ClickHouseClient;
+  private queryId: string;
+  private stream: StreamReadable<Row<unknown, "JSONCompactEachRow">[]>;
+  private rows: any[][] = [];
+
+  constructor(options: CursorOptions) {
+    super(options.chunkSize);
+    this.options = options;
+    this.client = options.client;
+    this.queryId = uuidv4();
+  }
+
+  async start(): Promise<void> {
+    log.info("Starting cursor");
+
+    const result = await this.client.query({
+      query: this.options.query,
+      query_params: this.options.params,
+      query_id: this.queryId,
+      format: "JSONCompactEachRow",
+    });
+
+    this.stream = result.stream();
+
+    return new Promise((resolve, reject) => {
+      this.stream.on("data", (rows) => {
+        rows.forEach((row) => {
```

Comment:
You need to `pause` the stream when you have `chunkSize` rows, otherwise you're not controlling how many rows get pushed to you from the client. What if it pushes a billion rows before `read` is called?

The ReadableStream implements `Stream` from the stream node library:

https://github.com/ClickHouse/clickhouse-js/blob/43751b0012e0a3e26523eed6ca333df7d81f7f4d/packages/client-node/src/result_set.ts#L23

Reply:
Ok, you really need to look at the `MysqlCursor`, as it implements this type of cursor exactly.

As it stands yours has a lot of bugs:

1. Your `start` promise doesn't resolve until the whole stream is empty, rather than when the cursor is ready for `read` to be called.
2. Your `read` method doesn't wait for the rows array to be `chunkSize` big
3. If the row buffer exceeds `chunkSize`, but `read` hasn't been called, you're not pausing the stream, you need to otherwise you get a massive memory leak

There are probably other problems, but that's def enough, lol. Def check the mysql implementation